hail/src/main/scala/is/hail/expr/ir/agg/StagedBlockLinkedList.scala

```diff
@@ -0,0 +1,236 @@
+    Code(setup,
+      firstNode := n,
+      lastNode := n,
+      totalCount := 0)
+  }
+
+  def init(r: Code[Region]): Code[Unit] =
+    initWithCapacity(r, defaultBlockCap)
+
+  private[agg] def pushNewBlockNode(r: Code[Region], cap: Code[Int]): Code[Unit] = {
+    val (setup, newNode) = allocateNode(r, cap)
+    Code(setup,
+      setNext(lastNode, newNode),
+      lastNode := newNode)
+  }
+
+  private[agg] def foreachNode(f: Node => Code[Unit]): Code[Unit] = {
+    Code(
+      tmpNode := firstNode,
+      p := true,
+      Code.whileLoop(p,
+        f(tmpNode),
+        p := hasNext(tmpNode),
+        tmpNode := next(tmpNode)))
+  }
+
+  def foreachElemAddress(f: EmitTriplet => Code[Unit]): Code[Unit] = {
+    foreachNode { n => Code(
+      i := 0,
+      Code.whileLoop(i < count(n),
+        f(elemAddress(n, i)),
+        i := i + 1))
+    }
+  }
+
+  def push(r: Code[Region], elt: EmitTriplet): Code[Unit] = {
+    val er = EmitRegion(mb, r)
+    Code(
+      (count(lastNode) >= capacity(lastNode)).orEmpty(
+        pushNewBlockNode(r, defaultBlockCap)), // push a new block if lastNode is full
+      elt.setup,
+      elt.m.mux(
+        pushMissing(lastNode),
+        pushPresent(lastNode, storeDeepCopy(er, elemType, elt.v, _))),
+      totalCount := totalCount + 1)
+  }
+
+  def pushFromAddress(r: Code[Region], addr: EmitTriplet): Code[Unit] =
+    push(r, EmitTriplet(addr.setup, addr.m,
+      Region.loadIRIntermediate(elemType)(addr.value)))
+
+  def append(r: Code[Region], bll: StagedBlockLinkedList): Code[Unit] = {
+    assert(bll.elemType.isOfType(elemType))
+    bll.foreachElemAddress(pushFromAddress(r, _))
+  }
+
+/*
+  def append(r: Code[Region], atyp: PArray, aoff: Code[Long]): Code[Unit] = {
+    assert(atyp.isOfType(bufferType))
+    //assert(!atyp.elementType.required)
+    val len = atyp.loadLength(region, aoff)
+    Code(
+      i := 0,
+      Code.whileLoop(i < len,
+        pushFromAddress(EmitTriplet(Code._empty,
+          atyp.isElementMissing(region, aoff, i),
+          atyp.elementOffset(aoff, len, i))),
+        i := i + 1))
+  }
+ */
+
+  def appendShallow(r: Code[Region], atyp: PArray, aoff: Code[Long]): Code[Unit] = {
+    assert(atyp.isOfType(bufferType))
+    assert(!atyp.elementType.required)
+    val len = atyp.loadLength(r, aoff)
+    Code(
+      tmpNode := r.allocate(nodeType.alignment, nodeType.byteSize),
+      initNode(tmpNode,
+        buf = aoff,
+        count = len),
+      setNext(lastNode, tmpNode),
+      lastNode := tmpNode,
+      totalCount := totalCount + len)
+  }
+
+  def writeToSRVB(srvb: StagedRegionValueBuilder): Code[Unit] = {
+    val er = EmitRegion(mb, srvb.region)
+    assert(srvb.typ.isOfType(bufferType))
+    Code(
+      srvb.start(totalCount, init = true),
+      foreachElemAddress { addr =>
+        Code(
+          addr.m.mux(
+            srvb.setMissing(),
+            deepCopy(er, elemType, addr.value, srvb.currentOffset)),
```

Comment:
`srvb.addWithDeepCopy` should do this, no?

Reply:
no, `StagedRegionValueBuilder` won't handle PString/PArray's correctly here. maybe there are certain functions to call that would produce the right kind of pointer but i couldn't figure out what that would look like

Reply:
right. I forgot that I fixed this in another PR somewhere, but it's not in master: #6830

Reply:
Not sure what you mean by not handling PStrings and PArrays correctly, though---those should have been ok.

Reply:
`addr.value` is a pointer to the pointer to the array data, but SRVB deepCopy expects just a pointer to the data

Reply:
your PR will clean this up though 😄